core/new-gui/src/app/workspace/service/autocomplete/model/autocomplete.service.ts

```diff
@@ -0,0 +1,175 @@
+   * a link is created between them, the attributed of the table selected in Source can be propagated to the KeywordSearch operator.
+   */
+  public invokeAutocompleteAPI(reloadCurrentOperatorSchema: boolean): void {
+    // get the current workflow graph
+    const workflowPlan = this.workflowActionService.getTexeraGraph();
+
+    // create a Logical Plan based on the workflow graph
+    const body = AutocompleteUtils.getLogicalPlanRequest(workflowPlan);
+    const requestURL = `${AppSettings.getApiEndpoint()}/${AUTOMATED_SCHEMA_PROPAGATION_ENDPOINT}`;
+
+    // make a http post request to the API endpoint with the logical plan object
+    this.httpClient.post<SuccessExecutionResult>(
+      requestURL,
+      JSON.stringify(body),
+      { headers: { 'Content-Type': 'application/json' } })
+      .subscribe(
+        // backend will either respond an execution result or an error will occur
+        // handle both cases
+        response => this.handleExecuteResult(response, reloadCurrentOperatorSchema),
+        errorResponse => this.handleExecuteError(errorResponse)
+      );
+  }
+
+  /**
+   * Handles valid execution result from the backend.
+   * Updates the current autocompleted schema. A value is inserted into 'autocompleteAPIExecutedStream' only
+   * if the current operator schema is to be reloaded (which is the case when a link is connected/deleted/disconnected)
+   * someplace in the workflow. However, if the operator's property is being changed, the autocomplete API is to be called,
+   * but the operator schema doesn't have to be reloaded.
+   *
+   * @param response
+   */
+  private handleExecuteResult(response: SuccessExecutionResult, reloadCurrentOperatorSchema: boolean): void {
+    this.operatorInputSchemaMap = response.result;
+    if (reloadCurrentOperatorSchema) {
+      this.autocompleteAPIExecutedStream.next('Autocomplete response success');
+    }
+  }
+
+    /**
+   * Handler function for unsuccessful schema propagation API call.
+   *
+   * Logs the error message to console.
+   *
+   * @param errorResponse
+   */
+  private handleExecuteError(errorResponse: HttpErrorResponse): void {
+    // error logged in console in different error scenarios
+    const displayedErrorMessage = AutocompleteUtils.processErrorResponse(errorResponse);
+    console.log(displayedErrorMessage.message);
+  }
+
+  /**
+   * returns the observable which produces array of source table names
+   */
+  private getSourceTablesNamesObservable(): Observable<ReadonlyArray<string>> {
+    return this.sourceTableNamesObservable;
+  }
+
+  /**
+   * Handles any kind of changes in the links of the joint graph and invokes the autocomplete API.
+   * There are 3 kinds of change streams exposed by joint graph wrapper - link add, link delete
+   * and link change.
+   */
+  private handleTexeraGraphLinkChangeEvent(): void {
+    merge(this.workflowActionService.getJointGraphWrapper().getJointLinkCellAddStream(),
```

Comment:
**!!! very important**: Don't use JointJS events, instead, listen to texera events. `this.workflowActionService.getTexeraGraph().getLinkAddStream()`, and `getLinkDeleteStream()`.

JointJS link events doesn't correspond directly to logical texera link add and delete events. They have very different semantics. The link change event is triggered continuously when you drag the link from one operator to another operator. The `invokeAutocompleteAPI` function will be called many many times and send many requests to the backend, and it also slows down the GUI.

JointJS link events correspond to the link in the UI. In order for you to fully understand the difference, please do: 
1. subscribe to each of the JointJS link (add, delete, change) event and print out messages in the console, so you can see yourself when do these events get triggered. 
2. do the same thing for texera link (add, delete) events and see when they get triggered. 
3. read the comments in `SyncTexeraModel.handleJointLinkEvents`

Reply:
Regarding the question about import `merge`, you can write `Observable.merge()`, (it's a static function bind to Observable) and then don't have to import it.